File: 01-mcp-concepts/mcp-resources/main_test.js
/** modulo de testing para probar el MCP */
const test = require('node:test');

const { Client } = require('@modelcontextprotocol/sdk/client/index.js');
const { StdioClientTransport } = require('@modelcontextprotocol/sdk/client/stdio.js');

// 1.- generamos el protocolo de transporte
const transport = new StdioClientTransport({
    command: 'node', // or node server.mjs | node server.cjs
    args: ['main.js']
});

// 2.- generamos el cliente
const client = new Client({
    name: 'calculator-client',
    version: '0.0.1'
});

test.describe('test mcp calculator', () => {     
    test.before(async () => {
        await client.connect(transport);
    });

    test.it('list tools', async t => {
        const {tools} = await client.listTools();
        t.assert.ok(
            Array.isArray(tools) && tools.length > 0, 
            'Tools is empty or not is array'
        );
    });
    
    test.it('add two numbers', async t => {
        const input = {
            a: 25.00,
            b: 25.00
        };

        const result = await client.callTool({
            name: 'add',
            arguments: input
        });
            
        const [{text}] = result.content;

        // verificamos los resultados
        t.assert.equal(text, String(input.a + input.b));
    })

    test.it('substract two numbers', async t => {
        const input = {
            a: 25.00,
            b: 10.00
        };

        const result = await client.callTool({
            name: 'substract',
            arguments: input
        });
            
        const [{text}] = result.content;

        // verificamos los resultados
        t.assert.equal(text, String(input.a - input.b));
    });

    test.it('multiply two numbers', async t => {
        const input = {
            a: 25.00,
            b: 5.00
        };

        const result = await client.callTool({
            name: 'multiply',
            arguments: input
        });
            
        const [{text}] = result.content;

        // verificamos los resultados
        t.assert.equal(text, String(input.a * input.b));
    });

    test.it('divide two numbers', async t => {
        const input = {
            a: 25.00,
            b: 5.00
        };

        const result = await client.callTool({
            name: 'divide',
            arguments: input
        });
            
        const [{text}] = result.content;

        // verificamos los resultados
        t.assert.equal(text, String(input.a / input.b));
    });

    test.it('divide by zero', async t => {
        const input = {
            a: 25.00,
            b: 0
        };

        try {
            const result = await client.callTool({
                name: 'divide',
                arguments: input
            });
            
        } catch (error) {
            t.assert.throws(() => {
                throw new Error('Error: cant divide by zero');
            });
        }
    });

    test.it('create resource', async t => {
        let name = 'gabriel'
        
        const result = await client.readResource({
            uri: 'greeting://' + name,
        });

        // verificamos los resultados del objeto
        t.assert.deepStrictEqual(result.contents[0], {
            uri: 'greeting://' + name,
            text: `Hello ${name}!`
        });
    });

    test.after(async () => {
        await client.close();
    });
});

File: 02-server-stdio/mcp-server-stio-go/main_test.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"os/exec"
	"testing"

	"github.com/modelcontextprotocol/go-sdk/mcp"
)

type Input struct {
	A float64 `json:"a"`
	B float64 `json:"b"`
}

func TestAdd(t *testing.T) {
	ctx := context.Background()
	input := Input{
		A: 25.00,
		B: 25.00,
	}

	// creamos el cliente mcp
	client := mcp.NewClient(&mcp.Implementation{Name: "test-mcp-server-stdio", Version: "0.0.1"}, nil)

	// ejecutamos el servidor mcp en segundo plano
	transport := &mcp.CommandTransport{Command: exec.Command("go", "run", "main.go")}
	session, err := client.Connect(ctx, transport, nil)

	// verificamos los errores
	if err != nil {
		log.Fatal(err)
	}

	defer session.Close()

	// ejecutamos la herramienta mcp
	params := &mcp.CallToolParams{
		Name:      "add",
		Arguments: input,
	}

	// evaluamos la expresion obtenida
	t.Run(fmt.Sprintf("%2.f+%2.f", input.A, input.B), func(t *testing.T) {
		expected := fmt.Sprintf("%.2f + %.2f = %.2f", input.A, input.B, (input.A + input.B))
		result, err := session.CallTool(ctx, params)

		// verificamos los errores
		if err != nil {
			t.Error(err)
			return
		}

		if result.IsError {
			t.Errorf("error: execute failed")
			return
		}

		var value string

		for _, content := range result.Content {
			value = content.(*mcp.TextContent).Text
		}

		if value != expected {
			t.Errorf("Add (%.2f + %.2f) = %s; expected %s", input.A, input.B, value, expected)
		}
	})
}

func TestMultiply(t *testing.T) {
	ctx := context.Background()
	input := Input{
		A: 25.00,
		B: 5.00,
	}

	// creamos el cliente mcp
	client := mcp.NewClient(&mcp.Implementation{Name: "test-mcp-server-stdio", Version: "0.0.1"}, nil)

	// ejecutamos el servidor mcp en segundo plano
	transport := &mcp.CommandTransport{Command: exec.Command("go", "run", "main.go")}
	session, err := client.Connect(ctx, transport, nil)

	// verificamos los errores
	if err != nil {
		log.Fatal(err)
	}

	defer session.Close()

	// ejecutamos la herramienta mcp
	params := &mcp.CallToolParams{
		Name:      "multiply",
		Arguments: input,
	}

	// evaluamos la expresion obtenida
	t.Run(fmt.Sprintf("%2.f*%2.f", input.A, input.B), func(t *testing.T) {
		expected := fmt.Sprintf("%.2f * %.2f = %.2f", input.A, input.B, (input.A * input.B))
		result, err := session.CallTool(ctx, params)

		// verificamos los errores
		if err != nil {
			t.Error(err)
			return
		}

		if result.IsError {
			t.Errorf("error: execute failed")
			return
		}

		var value string

		for _, content := range result.Content {
			value = content.(*mcp.TextContent).Text
		}

		if value != expected {
			t.Errorf("multiply (%.2f * %.2f) = %s; expected %s", input.A, input.B, value, expected)
		}
	})
}

func TestServerInfo(t *testing.T) {
	ctx := context.Background()

	// creamos el cliente mcp
	client := mcp.NewClient(&mcp.Implementation{Name: "test-mcp-server-stdio", Version: "0.0.1"}, nil)

	// ejecutamos el servidor mcp en segundo plano
	transport := &mcp.CommandTransport{Command: exec.Command("go", "run", "main.go")}
	session, err := client.Connect(ctx, transport, nil)

	// verificamos los errores
	if err != nil {
		log.Fatal(err)
	}

	defer session.Close()

	// ejecutamos la herramienta mcp
	params := &mcp.CallToolParams{
		Name:      "get_server_info",
		Arguments: nil,
	}

	t.Run("get_server_info", func(t *testing.T) {
		expected, err := json.Marshal(ServerInfoSchema{
			ServerName:   "example-stdio-server",
			Version:      "0.0.1",
			Transport:    "stdio",
			Capabilities: []string{"tools"},
			Description:  "Example MCP server using stdio transport (MCP 2025-06-18 specification)",
		})

		// verificamos los errores
		if err != nil {
			t.Error(err)
			return
		}

		result, err := session.CallTool(ctx, params)

		// verificamos los errores
		if err != nil {
			t.Error(err)
			return
		}

		if result.IsError {
			t.Errorf("error: execute failed")
			return
		}

		var value string

		for _, content := range result.Content {
			value = content.(*mcp.TextContent).Text
		}

		if value != string(expected) {
			t.Errorf("Error processing %v; expected %s", value, string(expected))
		}
	})
}

func TestGreeting(t *testing.T) {
	ctx := context.Background()

	// creamos el cliente mcp
	client := mcp.NewClient(&mcp.Implementation{Name: "test-mcp-server-stdio", Version: "0.0.1"}, nil)

	// ejecutamos el servidor mcp en segundo plano
	transport := &mcp.CommandTransport{Command: exec.Command("go", "run", "main.go")}
	session, err := client.Connect(ctx, transport, nil)

	// verificamos los errores
	if err != nil {
		log.Fatal(err)
	}

	defer session.Close()

	// ejecutamos la herramienta mcp
	params := &mcp.CallToolParams{
		Name: "get_greeting",
		Arguments: struct {
			Name string `json:"name"`
		}{Name: "Gabriel"},
	}

	t.Run("get_greeting", func(t *testing.T) {
		expected := fmt.Sprintf(
			"Hello %s! Welcome to the MCP stdio server.",
			"Gabriel",
		)

		result, err := session.CallTool(ctx, params)

		// verificamos los errores
		if err != nil {
			t.Error(err)
			return
		}

		if result.IsError {
			t.Errorf("error: execute failed")
			return
		}

		var value string

		for _, content := range result.Content {
			value = content.(*mcp.TextContent).Text
		}

		if value != expected {
			t.Errorf("Error processing %v; expected %s", value, expected)
		}
	})
}
